core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java

```diff
@@ -438,18 +438,33 @@ public boolean hasAliases(final String[] aliases, String[] concreteIndices) {
 
     /**
      * Returns indexing routing for the given index.
+     * If routing is not null then the value is returned, parent routing and alias routing are ignored.
+     * If routing is null and there is an alias routing then the value is returned and parent is ignored.
+     * If routing is null and a parent is defined then the parent is returned (unless an alias routing is defined, see above).
      */
     // TODO: This can be moved to IndexNameExpressionResolver too, but this means that we will support wildcards and other expressions
     // in the index,bulk,update and delete apis.
-    public String resolveIndexRouting(@Nullable String routing, String aliasOrIndex) {
-        if (aliasOrIndex == null) {
+    public String resolveIndexRouting(@Nullable String parent, @Nullable String routing, String aliasOrIndex) {
+        if (routing != null) {
+            // The routing is explicit, ignore the alias and parent routing (if any).
```

Comment:
@rjernst that's how it was done before but the purpose of this PR is to prioritize an explicit routing (set by the user) over an alias routing and a parent routing. IMO if the user adds a routing to the request we should always honor it and ignore the rest of the rules (parent and alias). Regarding the users confusion when they use the API I don't think the current situation is better (see #3068). I can restore the current behavior and only apply the parent/alias priority proposed in #3068 if you want but I have the feeling that it would bring more confusion. Bottom line is that if the user sets a routing then we should trust that he knows what he's doing ;).